src/simulators/densitymatrix/densitymatrix_state.hpp

```diff
@@ -0,0 +1,725 @@
+
+  //-----------------------------------------------------------------------
+  // Measurement Helpers
+  //-----------------------------------------------------------------------
+
+  // Return vector of measure probabilities for specified qubits
+  // If a state subclass supports this function it then "measure"
+  // should be contained in the set returned by the 'allowed_ops'
+  // method.
+  // TODO: move to private (no longer part of base class)
+  rvector_t measure_probs(const reg_t &qubits) const;
+
+  // Sample the measurement outcome for qubits
+  // return a pair (m, p) of the outcome m, and its corresponding
+  // probability p.
+  // Outcome is given as an int: Eg for two-qubits {q0, q1} we have
+  // 0 -> |q1 = 0, q0 = 0> state
+  // 1 -> |q1 = 0, q0 = 1> state
+  // 2 -> |q1 = 1, q0 = 0> state
+  // 3 -> |q1 = 1, q0 = 1> state
+  std::pair<uint_t, double>
+  sample_measure_with_prob(const reg_t &qubits, RngEngine &rng);
+
+
+  void measure_reset_update(const std::vector<uint_t> &qubits,
+                            const uint_t final_state,
+                            const uint_t meas_state,
+                            const double meas_prob);
+
+  //-----------------------------------------------------------------------
+  // Special snapshot types
+  //
+  // IMPORTANT: These methods are not marked const to allow modifying state
+  // during snapshot, but after the snapshot is applied the simulator
+  // should be left in the pre-snapshot state.
+  //-----------------------------------------------------------------------
+
+  // Snapshot current qubit probabilities for a measurement (average)
+  void snapshot_probabilities(const Operations::Op &op,
+                              OutputData &data,
+                              bool variance);
+
+  // Snapshot the expectation value of a Pauli operator
+  void snapshot_pauli_expval(const Operations::Op &op,
+                             OutputData &data,
+                             bool variance);
+
+  // Snapshot the expectation value of a matrix operator
+  void snapshot_matrix_expval(const Operations::Op &op,
+                              OutputData &data,
+                              bool variance);
+
+  //-----------------------------------------------------------------------
+  // Single-qubit gate helpers
+  //-----------------------------------------------------------------------
+
+  // Apply a waltz gate specified by parameters u3(theta, phi, lambda)
+  void apply_gate_u3(const uint_t qubit, const double theta, const double phi,
+                     const double lambda);
+
+  // Optimize phase gate with diagonal [1, phase]
+  void apply_gate_phase(const uint_t qubit, const complex_t phase);
+
+  //-----------------------------------------------------------------------
+  // Config Settings
+  //-----------------------------------------------------------------------
+
+  // OpenMP qubit threshold
+  int omp_qubit_threshold_ = 14;
```

Comment:
This is actually referring to the number of qubits in the underlying QubitVector (which is twice the number of qubits for the density matrix).

At some point it would be good to subclass the State classes for qubitvector and its derived ones like unitary/density matrix so stuff like this is less confusing